core/src/main/java/org/infinispan/interceptors/locking/PessimisticLockingInterceptor.java

```diff
@@ -1,126 +1,94 @@
- * Design note: when a lock "k" needs to be acquired (e.g. cache.put("k", "v")), if the lock owner is the local node,
- * no remote call is performed to migrate locking logic to the other (numOwners - 1) lock owners. This is a good
- * optimisation for  in-vm transactions: if the local node crashes before prepare then the replicated lock information
- * would be useless as the tx is rolled back. OTOH for remote hotrod/transactions this additional RPC makes sense because
- * there's no such thing as transaction originator node, so this might become a configuration option when HotRod tx are
- * in place.
- *
- * Implementation note: current implementation acquires locks remotely first and then locally. This is required
- * by the deadlock detection logic, but might not be optimal: acquiring locks locally first might help to fail fast the
- * in the case of keys being locked.
+ * Locking interceptor to be used by pessimistic transactions.
+ * <p>
+ * Design (from 9.1):
+ * <p>
+ * This interceptor is skipped when {@link Flag#SKIP_LOCKING} is used or when all affected keys are already locked.
+ * Read-only commands can acquire the locks when {@link Flag#FORCE_WRITE_LOCK}.
+ * <p>
+ * The new design has 2 concepts. The primary lock and the backup lock. The primary lock is acquired in the primary
+ * owner and it is exclusive; only one transaction can acquire it. The backup lock is acquired by the backup owner and
+ * it is shared; multiple transactions can acquire simultaneously. The backup locks are associated with the topology id
+ * and they are only used when the primary owner leaves or crashes.
+ * <p>
+ * This algorithm contacts all owners and acquires the primary lock in the primary owner and the backup lock in the
+ * backup owners. So, lock acquisition (single or multiple keys) happens in a single RPC to all the owners of the
+ * key(s). This is triggered in {@link TxDistributionInterceptor#visitLockControlCommand(TxInvocationContext,
+ * LockControlCommand)}.
+ * <p>
+ * The {@link Flag#CACHE_MODE_LOCAL} changes this behaviour; it makes the lock acquisition only local. So, only local
+ * the primary (if primary owner) or backup lock (if backup owner) is acquired. No RPC is create and if the node isn't
+ * an owner, no locks are acquired.
+ * <p>
+ * <p>
+ * In case of primary owner changes, there are 2 cases.
+ * <p>
+ * 1) a new node joins and it is the new primary owner. In this case, the old primary owner transfers the lock
+ * information to the new primary owner.
+ * <p>
+ * 2) the primary owner crashes or leaves. In this case, a backup owner will be promoted to the primary owner. All
+ * primary lock requests are hold until its previous backup locks are released. Note that a transaction only acquires
+ * the lock once. If a transaction requests a primary lock but has the backup lock acquired, it doesn't grant the
+ * primary lock if there are other backup locks from other transactions.
```

Comment:
Shouldn't this be "If a transaction has acquired the backup lock, it will not try to acquire the primary lock" ?